### ISSUE TITLE
Reassociate every homogeneous recursion: the squaring that bought 3.57× on the Potts chain, applied repo-wide

### ISSUE THREAD

**Issue by michaelJwilson** [michaelJwilson/snakes_and_ladders#785]
**TL;DR:** #775 found that `hmc.sample`'s two top terms were one loop, the per-site transfer recursion (`logsumexp`, 33.2%) and its tape (`run_backward`, 41.7%), and that reassociating the homogeneous product `T^(L-1)` by repeated squaring cut both: 23.5 s → 6.6 s, 3.57×, `logsumexp` 512,000 calls → 96,000. The same shape recurs across the package. This ticket is the abstract opportunity: find every recursion that applies one associative operator `L` times, classify it, measure it, and reassociate where the bar is met.

### The profile
- The cost of a taped recursion is two-fold and both scale with `L`: the work, `L` applications of the step, and the tape, `L` autograd nodes whose backward is paid per node. On #775's instance the tape was the larger term (41.7% against 33.2%).
- Two shapes of recursion, and each has a reassociation: **(A) homogeneous**, the same operator at every step, `T^L v`, which exponentiation by squaring brings to `O(log L)` work and `O(log L)` tape; **(B) inhomogeneous but associative**, `A_L ... A_1 v` with a per-step factor, where a pairwise (scan) reassociation keeps the work at `O(L)` and brings the tape to `O(log L)` depth. (A) cuts both terms; (B) cuts the tape alone, so it is worth measuring where the tape dominates.
- Candidates read from the site loops: (A) the chain marginals in `opt/potts.py` lines 184 and 190 and `sim/potts.py` lines 619 and 627, prefix and suffix products of one `T`; the lattice transfer product in `PottsLatticeObjective` where the field is uniform. (B) `opt/hmm.py` line 936, the forward algorithm with a per-site emission; `likelihood/forward_backward.py`, whose kernel varies per step since #653; `search/gibbs.py` lines 826 and 833, the chain messages of the factor graph. Not candidates: the tree pruning, where `torch.matrix_exp` already squares per branch; the simulators, which draw sequentially; `learn/exact.py`, which enumerates.

### Route
Pure Python and PyTorch, as #775: a reassociation is an algorithmic cut, which root `CLAUDE.md` ranks above a mechanical one, and it lands before any port. `squaring_is_cheaper` (#775) is the rule for (A) and gains the (B) counterpart, so each site keeps its recursion as the oracle and chooses by the rule, never silently.

### Sizes
The stress instance of each family: the chain of 64 at 1,000 draws (#775), the HMM and ragged-HMM stress fixtures, the factor-graph chain at its stress length. A ratio read at a gate size decides nothing.

### Oracle and tolerance
The recursion each site already runs, at the tolerance its existing test declares; bitwise is the target and #775 reached it on the chain, but a reassociation reorders a reduction and may back off to the declared tolerance with the difference stated.

### The bar and the revert
Root `CLAUDE.md`'s: at least 2× on the enclosing call at stress size, with the absolute saving stated against that call; a candidate that reads under it is recorded in the experiment file and left as it is.

### Prior measurements
#775: `hmc.sample` 23.501 / 23.597 s → 6.601 / 6.511 s, 3.57× (3.60× per gradient), saving 17.0 s of 23.5; `torch.logsumexp` 512,000 calls / 8.19 s → 96,000 / 1.69 s. #653: the varying kernel in `forward_backward` is the reason (B) applies there and (A) does not.

### How it will be validated
- One experiment file per candidate measured, both walls, and the survey table (site, shape, `L` at stress, tape share, ratio, verdict) in `STATUS.md`.
- `-m critical` per PR; the candidate's `oracle` tests at gate and stress sizes; the full non-release tier at the top of the stack. Select: `uv run --no-sync pytest tests/regression/opt/test_opt_potts.py tests/regression/opt/test_opt_hmm.py tests/regression/likelihood/test_forward_backward.py tests/regression/search/test_gibbs.py -n 3 -m "not release"`.
- One PR per site family, each body at most 40 lines; after #775 merges.
